modules/history/src/history-helper.js

```diff
@@ -0,0 +1,186 @@
+const SerializableClass = require("abitbol-serializable");
+const self = require("../index.js");
+
+const historyHelper = {
+
+    /**
+     * Check the differences between two state and apply the changes of snapshotState
+     * to currentState
+     * @param  {Object} currentState - The current state of the history
+     * @param  {Object} snapshotState - The new state of the history
+     */
+    applySnapshotDifference: (currentState, snapshotState) => { // eslint-disable-line
+        const {
+            dataStore,
+        } = self.app.modules;
+        // We make a flat object list for each object
+        const currentEntities = historyHelper.flattenSnapshot(currentState);
+        const snapshotEntities = historyHelper.flattenSnapshot(snapshotState);
+
+        Object.keys(currentEntities).forEach(
+            (currentEntityKey) => {
+                let entityFound = false;
+                Object.keys(snapshotEntities).forEach(
+                    (snapshotEntityKey) => {
+                        // Check common entity properties
+                        const currentEntity = currentEntities[currentEntityKey];
+                        const snapshotEntity = snapshotEntities[snapshotEntityKey];
+                        if (currentEntityKey === snapshotEntityKey) {
+                            entityFound = true;
+                            const entity = dataStore.getEntity(currentEntityKey);
+                            // Property is different or doesn't exist
+                            const unserializedEntity = SerializableClass
+                                .$unserialize(snapshotEntity);
+                            Object.keys(snapshotEntity).forEach(
+                                (property) => {
+                                    if ((typeof currentEntity[property] !== "undefined"
+                                        && historyHelper.checkDiff(currentEntity[property],
+                                            snapshotEntity[property]))
+                                        || !currentEntity[property]) {
+                                        // Found differences
+                                        entity[property] = unserializedEntity[property];
+                                    }
+                                },
+                            );
+                            // Check if we don't need to remove a property in entities
+                            // Should never happen
+                            Object.keys(currentEntity).forEach(
+                                (property) => {
+                                    if (typeof snapshotEntity[property] === "undefined") {
+                                        self.app.log.warn("You are removing a property between two entities, which should never happen");
+                                        delete entity[property];
```

Comment:
Why this should never happen ? 
Can't I set `myEntity[property] = undefined` and then apply a snapshot ?

There seems to be a confusion between those cases :
- snapshotEntity has a property which value is "undefined"
- snapshotEntity has no property

Reply:
Currently the rules is that when you compare two object there should not be a property that doesn't exist between two object, if it's the case it means that there something wrong in how you defined your structure or how you manage it.

Reply:
Example: `const foo = {bar: undefined}`
- Property "bar" **does** exists, since `Object.keys(foo).includes("bar")`
- However `typeof foo["bar"] === undefined`

In my opinion, `typeof foo["bar"] === undefined` conditions should be replaced by `!Object.keys(foo).includes("bar")`

Reply:
As `undefined` is not a valid JSON value, we can't serialize it so I guess it ok to confuse those two cases